src/Agent/Communication/AbstractRequest.php

```diff
@@ -148,8 +148,7 @@ protected function setMode($mode)
     */
     private function guessMode($contents)
     {
-        json_decode($contents);
-        if (json_last_error() === JSON_ERROR_NONE) {
+        if ($this->headers->hasHeader('GLPI-Agent-ID') && preg_match('/^{/', $contents)) {
```

Comment:
I guess we can use header here, but we can also keep switching to json if the content is a real json (`json_last_error() === JSON_ERROR_NONE`)
```suggestion
        json_decode($contents);
        if ((json_last_error() === JSON_ERROR_NONE || $this->headers->hasHeader('GLPI-Agent-ID')) && preg_match('/^{/', $contents)) {
```
I'm not sure I understand the regexp part, maybe just add a comment

Reply:
Here, `$contents` should only be XML or JSON content. Any agent trying to communicate setting `GLPI-Agent-ID` header can communicate in XML or JSON. JSON format MUST start with `{` and XML with `<` afaik.
I uses a simple preg_match on first char here but we can also just test the first char equality with `{` if you think this is better. Anyway this is a far more quicker handling than decoding `$contents` eventually expecting a JSON: even if this is quick, decoding a big JSON inventory is resource consuming. Anyway from here we just have to assume the content is JSON or XML to route the request handle to `handleXMLRequest()` or `handleJSONRequest()`. If we route to `handleJSONRequest()`, it will always have to decode the JSON, so it's better to avoid a double decoding here.

Reply:
Well, there is no need to test for header then:
```suggestion
        if (preg_match('/^{/', $contents)) {
```
Should be enough to guess the mode. It seems simplier for tests or "mocks" to just pass JSON data and not a HTTP header.
JSON validity may be checked in `handleJSONRequest()` like XML one is done in `handleXMLRequest`, yes.

Reply:
Yes, this is enough to guess the content is JSON or XML, but `GLPI-Agent-ID` is a requirement for JSON protocol so it's an error to switch to JSON mode without this header. Not providing `GLPI-Agent-ID` should result in a "XML not well formed!" 400 error even if the agent provides a JSON as we should expect XML content without such header.
For tests, you simply have to call `$inventory_request->handleContentType('application/json');` when testing with JSON as done to fix TU in 5c728de.
I also added a test where I pass `GLPI-Agent-ID` to Guzzle client, I don't think this is a big deal.

Reply:
Yes, I agree, supporting `Content-Encoding` should be better as it could simplify the `Content-Type` support and even it could help GLPI with no need to guess the mode to use. I'll investigate `Content-Encoding` support in GLPI-Agent and GLPI for another PR.

Reply:
Two points on this part

- To discuss again about the preg_match, i guess it's to detect the content is in json
   The best way to do that in php si to call `json_decode` and check `json_last_error()` like done below in this pr.
   The preg_match may be slower (i didn't check) but it's clearly harder to read.
   But i can live with current code.

   But i agree with Johan, a comment could clarify you want to guess it's a json content (by detecting first caracter is `{ if you stay with the `preg_match`.

- The check on agent header seems superflu for the purpose of the method (guessing content format) and if it's needed maybe moved to another method validating the request. 
   We try to check with a content parameter and the method implies to check a property of the current objet.

   But more generally on the exchange, do the header `GLPI-Agent-ID` is mandatory to send an inventory json file ?
   We talked tuesday about various plugins sending inventory data (like sccm or nutanix) and they may work without agent (I'm note sure about today with their current integrations).
  Maybe i'm not sufficienctly aware of the whole code but i need explanantion on this.
  In the case of an inventory plugin, do we need to declare a virtual agent to send inventory data ?

Reply:
I understand the first point now. Just checking `$contents` first char is definitively better to decide if it is a JSON or XML content.
About `GLPI-Agent-ID`, it is mandatory in spec, but you're right, this is not the right place to make this check. In my head, I though that check was a good idea to decide if we need to run the `preg_match` or not but just checking `$contents` first char make this test even less optimum.
So I modified the code to be simpler and just focus on the purpose in 125d382.